asr1k_neutron_l3/plugins/l3/service_plugins/l3_extension_adapter.py

```diff
@@ -520,6 +548,23 @@ def _find_ips_for_dynamic_nat_pool(self, context, subnet_id, ip_count):
 
         return None, None
 
+    def _find_gateway_ip_for_dynamic_nat_pool(self, context, subnet_id, nat_pool_ips):
+        """Find a usable gateway ip for a nat pool in a subnet without fragmenting the ip space too much"""
+        # current allocation logic: last ip that is not allocated and not in the pool
+        allocation_pools, ip_allocations = self._get_subnet_allocation_pools_and_allocations(context, subnet_id)
+
+        gw_ip_candidate = None
+        for ip_pool in allocation_pools:
+            for ip in netaddr.iter_iprange(ip_pool.last_ip, ip_pool.first_ip, step=-1):
+                # don't need to look at smaller ips if we have a candidate
+                if gw_ip_candidate and ip < gw_ip_candidate:
+                    break
+
+                if ip not in ip_allocations and ip not in nat_pool_ips:
+                    gw_ip_candidate = ip
```

Comment:
```suggestion
                    gw_ip_candidate = ip
                    break
```
nit: I think we could also continue with the next allocation_pool here, since we are iterating in reverse order over all IPs, so we know that we would break in the next iteration on line 561 either way.

Reply:
That's true. I thought what would be nicer and opted for "just one `break`" but the other option felt equally nice to me at time of writing. As I like both, what do the reviewers vote? ;)

Reply:
No strong opinions (hence also the "nit") so I leave that decision to you. ^^

Reply:
In general I prefer to break/return/terminate as early as possible. When reading code, especially my own, I would probably be a bit confused and try to find out why the search was continued although it apparently already found a candidate. Maybe there is a side effect, maybe its needed, but why? what am I missing?

Reply:
We might have multiple allocation pools (we nearly never have that but still have to handle the case) and the second pool might be larger than the first. I'll add the extra break then. :)